### PR TITLE
TS Feature Analysis and Plots

### PR DESCRIPTION
**Pull request by alexdkellogg**
Submitting pull request for a set of functions that plot the time series, counterfactually predicted time series, and treatment effects over time.

### CODE REVIEW

File: r_code/ts_feature_analysis.R [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+TSFeaturesPCScatterplot <- function(data_full,
+                                    id_var = "entry",
+                                    outcome_var = "target",
+                                    treat_indicator = "treatperiod_0",
+                                    type=c("rug", "centroid", "contour"),
+                                    ...) {
+  if (missing(type)) {
+    type <- "rug"
+  } else {
+    type <- match.arg(type)
```
[the remark this answers is not in the document]
Comment (alexdkellogg) [reply r_code/ts_feature_analysis.R:32]: I believe you're right, I'll remove the redundant code.

File: r_code/ts_feature_analysis.R [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+                   axis.line = element_line(colour = "black"))
+  # Add to the plot based on the input type.
+  if(type=="rug"){
+    feature_overlap_plot=feature_overlap_plot+ 
+      ggplot2::geom_rug(aes(color=Treated)) 
+  } else if(type=="centroid"){
+    feature_overlap_plot <- .CentroidHelper(feature_overlap_plot, pc_entry_df)
+  } else if(type=="contour"){
+    feature_overlap_plot <- feature_overlap_plot +
+      ggplot2::geom_density_2d(aes(color = Treated,alpha=..level..), size=2)
```
[the remark this answers is not in the document]
Comment (alexdkellogg) [reply r_code/ts_feature_analysis.R:72]: https://stackoverflow.com/questions/32206623/what-does-level-mean-in-ggplotstat-density2d

The second answer I found useful. I believe the geom_density_2d function is creating a new data frame with a column named "level", and this is the way it is typically references when used for the plots. So I believe it's an automated way to grab the percentiles from this newly created data frame.

File: r_code/ts_feature_analysis.R [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+      mean_pc2 = mean(PC2)
+    )
+  
+  # Compute the difference between the two centroids.
+  centroid_dist <- pc_centroids %>%
+    tidyr::pivot_wider(names_from = Treated, 
+                       values_from = c(mean_pc1, mean_pc2)) %>%
+    dplyr::mutate(
+      dist =
+        (mean_pc1_0 - mean_pc1_1)^2 + (mean_pc2_0 - mean_pc2_1)^2
```
[the remark this answers is not in the document]
Comment (alexdkellogg) [reply r_code/ts_feature_analysis.R:123]: The scatterplot is such that the x-axis is PC1 and the y-axis is PC2. So really mean_pc1_0=x_control, mean_pc1_1=x_treated, and similar for mean_pc2_Z. That is, the number after PC refers to the axis, and the final number refers to group status (which identifies the point in our 2d plot). I checked again and I believe the code does this correctly, but happy to follow up as I could be wrong!

File: r_code/ts_feature_analysis.R [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+  } else {
+    type <- match.arg(type)
+  }
+  # Identify treated observations.
+  treated_entries <- (data_full %>%
+                        dplyr::filter(!!as.name(treat_indicator) > 0) %>%
+                        dplyr::distinct(!!as.name(id_var)) %>%
+                        dplyr::pull())
+  data_full <- data_full %>%
+    dplyr::mutate(Treated = as.numeric(!!as.name(id_var) %in% treated_entries))
```
[the remark this answers is not in the document]
Comment (alexdkellogg) [reply r_code/ts_feature_analysis.R:40]: Great suggestion.

File: r_code/ts_feature_analysis.R [hunk 5]
```diff
@@ -11,41 +9,30 @@ pacman::p_load(dplyr, ggplot2, janitor,tsfeatures,tibble, tidyr,rstatix)
...
+  type <- match.arg(type)
   # Identify treated observations.
-  treated_entries <- (data_full %>%
-                        dplyr::filter(!!as.name(treat_indicator) > 0) %>%
-                        dplyr::distinct(!!as.name(id_var)) %>%
-                        dplyr::pull())
-  data_full <- data_full %>%
-    dplyr::mutate(Treated = as.numeric(!!as.name(id_var) %in% treated_entries))
+  data_full <- .IdentifyTreated(data_full=data_full, id_var=id_var,
+                                            treat_indicator=treat_indicator)
```
Comment (jeffreyoldham) [inline r_code/ts_feature_analysis.R:28]:
```suggestion
  data_full <- .IdentifyTreated(data_full=data_full, id_var=id_var,
                                                   treat_indicator=treat_indicator)
```

File: r_code/ts_feature_analysis.R [hunk 6]
```diff
@@ -209,15 +183,31 @@ FeatureDensity <- function(data_full,id_var = "entry",
...
   return(feature_density)
   
 }
+
+#' Identify and tag the treated units by ID.
+#'
+#' @inheritParams TSFeaturesPCScatterplot
+#'
+#' @return The original tibble of data, with a new Treated indicator column. 
+.IdentifyTreated <- function(data_full, id_var, treat_indicator){
```
Comment (jeffreyoldham) [inline r_code/ts_feature_analysis.R:204]:
```suggestion
.IdentifyTreated <- function(data_full, id_var, treat_indicator) {
```

File: r_code/ts_feature_analysis.R [hunk 7]
```diff
@@ -209,15 +183,31 @@ FeatureDensity <- function(data_full,id_var = "entry",
...
+  # Identify treated entries.
+  treated_entries <- (data_full %>%
+                        dplyr::filter(!!as.name(treat_indicator) > 0) %>%
+                        dplyr::distinct(!!as.name(id_var)) %>% 
+                        dplyr::pull())
```
Comment (jeffreyoldham) [inline r_code/ts_feature_analysis.R:209]:
```suggestion
                      dplyr::filter(!!as.name(treat_indicator) > 0) %>%
                      dplyr::distinct(!!as.name(id_var)) %>% 
                      dplyr::pull())
```

Comment (alexdkellogg) [reply r_code/ts_feature_analysis.R:209]: I think the parenthesis at the beginning and end were messing up the default alignment for piping. It's a different change than requested, but it's what the other lines/files have used.

**Pull request merged**